### DISCUSSION TITLE
How to accumulate `packages.bib` used in a Quarto book

### DISCUSSION THREAD

**Discussion by friendly** [quarto-dev/quarto-cli#5806] [Q&A]
### Description

# How to accumulate `packages.bib` used in a Quarto book

I'm writing a book with multiple chapters and wish to generate a `packages.bib` file at
the end containing citations for all the packages that were loaded via `library()`
while the book is compiled by "Render book". I know how to do this for a vignette, but not
for a multi-chapter quarto book. If this is currently difficult in Quarto, solving this would be a great **New Feature**

In vignettes I use the following to automatically generate a `packages.bib` file for all
the packages used, so that I can easily cite them as needed:

* In the preamble I list a few packages which may be referred to in the text, but not
actually loaded by `library()`

```
# packages to be cited here. Code at the end automatically updates packages.bib
.to.cite <- c("nnet",  "geomtextpath")
```

* Then, at the end, just before `## References` I use this chunk to join `.to.cite` with
`.packages()` and write this out using `knitr::write_bib()`:

````
```{r write-bib, echo = FALSE}
# write a packages.bib file of the packages (.packages()) that have been used here
pkgs <- unique(c(.to.cite, .packages()))
knitr::write_bib(pkgs, file = here::here("vignettes", "packages.bib"))
```
````

**Comment by mcanouil** [comment]
If Quarto implements such feature, it would also need to do so for all the other languages (very unlikely).

If you use explicit library/require, you could simply leverage `renv::dependencies()` to list your R dependencies, then write the bib file.

See <https://rstudio.github.io/renv/reference/dependencies.html>.

All this could be implemented as an R script used in pre-render project scripts, see <https://quarto.org/docs/projects/scripts.html#pre-and-post-render>.

**Comment by friendly** [comment]
Thanks for the quick reply. I hadn't thought of `renv`.  

This could make a fabulous, concrete example of how to use pre-post-render in a project!  This topic (like many others in Quarto docs) is sorely lacking useful examples.

**Reply by mcanouil** [reply]
The thing is that this is an R example.
Quarto is language agnostic.
If the documentation starts to have example for any supported languages through knitr/Jupyter, users won't see Quarto's features.
Pre/Post render scripts can do absolutely anything.

Your question has nearly nothing to do with Quarto, as it is also valid for R scripts and RMarkdown scripts.

This being said, this can easily shown in a blog post by any community member.

PS: I personally think this question is more appropriate in the "Q&A" category rather than "Feature Requests".

**Comment by friendly** [comment]
Frankly, I don't think this is a stance Quarto should publicize.
IMHO, language-agnostic documentation is often too abstract for users to see how to apply it to their language-bound applications. Writing good documentation is hard, but the user task is made easier by seeing concrete uses that solve problems.
To say that "_Pre/Post render scripts can do absolutely anything_" is cold comfort, and discouraging to a new Quarto learner. 

Still, I'll move discussion of this particular problem to RStudio discussion

Point taken about Q&A vs. Feature Request

**Reply by mcanouil** [reply]
I understand, the question about showing various examples is tricky.
There are some discussions about this.

For instance, the best way to show pre/post render scripts in action is not really possible in documentation without cluttering the documentation and in the end, it won't tell the user how to do what they want in a script language of their choosing.
So probably that an example gallery is better suited than adding a bunch of examples with no real use case.

The documentation shows an example of using four types of files as pre-render and even a make command, see <https://quarto.org/docs/projects/scripts.html#pre-and-post-render>.
It's regular TypeScript, Python, R or LUA code.
So users can easily choose the language they are mastering the most and do what they want to do in that language, then simply pass that to "pre-render".

**Reply by mcanouil** [reply]
If you have a concrete example to improve the documentation, please do.
It would be very appreciated.

**Reply by mcanouil** [reply]
Here an example: https://github.com/mcanouil/mickael.canouil.fr

- https://github.com/mcanouil/mickael.canouil.fr/blob/main/_quarto.yml
- https://github.com/mcanouil/mickael.canouil.fr/blob/main/publications.R

As you can see, it's a 100 % regular R script.

**Comment by friendly** [comment]
Thank you so much for these very helpful suggestions.

I agree that an example gallery would solve both my (and others) problems and yours -- it would be easy to just link to these in the documentation. User-contributed examples would broaden the scope.  Your Awesome-Quarto could serve as a hub for this.

**Reply by mcanouil** [reply]
I forgot to mention that there are also various examples in GitHub Discussions.
So using some keywords can narrow down the list quite efficiently to get examples based on users previous questions.
